### PR TITLE
feat(evaluation): log eval-{config_id} W&B artifact with R2 references

### LINKED ISSUE #1473: feat(evaluation): log eval-results W&B artifact with R2 references
## Summary

Log an `eval-results` W&B artifact (name `eval-{eval_config_id}`, type `eval-results`) from the eval entry point, with `add_reference` to the R2 outputs (predictions, audio, metrics). Per `storage-provenance-spec.md` §4–5.

## Current state

`cli/eval.py:312` (`_maybe_upload_output_dir`) uploads the eval output dir to R2 but logs no W&B artifact, so eval results are not indexed in W&B and cannot anchor the lineage tail.

## Deliverables

- [ ] Create `wandb.Artifact(name=f"eval-{eval_config_id}", type="eval-results")`
- [ ] `add_reference` to the R2 eval output prefix
- [ ] `artifact.metadata` carries summary metrics + git_sha
- [ ] Tests assert name/type/references (mocked W&B)

## Dependency

Tail of the lineage DAG. Depends on #403 (run IDs); consumes model + dataset artifacts (see lineage task). Refs #93, #1470, #1467.

### PR DESCRIPTION
**Pull request by ktinubu**
## Summary

Logs an `eval-results` W&B artifact (name `eval-{eval_config_id}`, type `eval-results`) from the eval entrypoint after the output dir is mirrored to R2, mirroring `finalize_dataset`'s dataset-artifact pattern. This anchors the lineage-DAG tail per `docs/design/storage-provenance-spec.md` sections 4-6.

## What changed

- `build_eval_results_artifact(cfg, output_dir_uri, metric_dict, git_sha)` — builds `wandb.Artifact(name=f"eval-{resolve_run_config_id(cfg)}", type="eval-results")`, `add_reference`s the R2 output prefix as an `s3://` URI (`checksum=False`, since R2's custom S3 endpoint is unreachable by W&B's reference handler), and records the scalar summary metrics plus `git_sha` in `metadata`.
- `_eval_summary_metrics` reduces the `metric_dict` to JSON-safe scalar floats so `artifact.metadata` stays small.
- `_log_eval_results_artifact` iterates `WandbLogger` entries best-effort: a wandb failure warns and is swallowed so artifact logging never aborts a completed eval. No-op without a `WandbLogger`, with a null upload URI, or off the global-zero rank.
- Wired into `main()` after `_maybe_upload_output_dir`, gated on `trainer.is_global_zero`.
- Adds a shared `r2_io.to_s3_uri` helper (alongside `to_rclone_path`) instead of duplicating finalize's private rewrite, and reuses `spec._get_git_sha` for the commit SHA.

## Testing

- New `tests/test_eval_artifact.py`: asserts name/type, the `s3://` reference equals the upload prefix, metadata round-trip, scalar coercion of tensors, and the no-op paths (no `WandbLogger`, null URI) plus error-swallow.
- New `TestToS3Uri` in `tests/pipeline/test_r2_io.py`.
- Full fast suite passes (1850 passed, 6 skipped; run serially due to host memory pressure — parallel xdist workers segfault under ~3 GB free, unrelated to this change).

Closes #1473
Refs #1467, #1470, #93

### CODE REVIEW

**Comment by ktinubu** [comment]
## Verification Results

All checks run against the PR branch (`feat/eval-results-wandb-artifact-1473`, HEAD 722ea81).

- [x] **The eval-results artifact has the spec'd name/type, an `s3://` reference to the R2 prefix, and round-tripped scalar metadata** *(Level 2 — invoked the real builder; the promise is W&B artifact construction, not an end-to-end run)*
  ```bash
  $ uv run python -c '
  from omegaconf import OmegaConf
  import torch
  from synth_setter.cli.eval import build_eval_results_artifact
  from synth_setter.pipeline import r2_io
  cfg = OmegaConf.create({"task_name": "nsynth-v1"})
  uri = "r2://eval-artifacts/eval-nsynth-v1-20260606T000000000Z"
  md = {"test/param_mse": torch.tensor(0.25), "vec": torch.tensor([1.0, 2.0]), "flag": True}
  art = build_eval_results_artifact(cfg, uri, md, "0"*40)
  print("name:", art.name); print("type:", art.type)
  print("refs:", {e.ref for e in art.manifest.entries.values()})
  print("metadata:", art.metadata)
  print("to_s3_uri:", r2_io.to_s3_uri(uri))
  '
  name: eval-nsynth-v1
  type: eval-results
  refs: {'s3://eval-artifacts/eval-nsynth-v1-20260606T000000000Z'}
  metadata: {'test/param_mse': 0.25, 'git_sha': '0000000000000000000000000000000000000000'}
  to_s3_uri: s3://eval-artifacts/eval-nsynth-v1-20260606T000000000Z
  ```
  Name `eval-{config_id}`, type `eval-results`, the reference equals the R2 prefix rewritten to `s3://`, and metadata carries the scalar metric + `git_sha` while dropping the vector and the `bool`.

- [x] **End-to-end: the wiring `main()` invokes writes a real `eval-results` artifact (name, type, `s3://` reference, metadata) through the live wandb client** *(Level 1 — drove `_log_eval_results_artifact` through a real `WandbLogger(offline=True)`, finished the run, and decoded the `run-*.wandb` binary; no wandb internals mocked)*
  ```bash
  $ E2E_TMP=$(mktemp -d) PYTHONPATH=. uv run python /tmp/verify_eval_e2e.py
  E2E PASS: eval-results artifact, type, s3 reference, and metadata all recorded
  ```
  The decoded binary contains `eval-nsynth-e2e`, `eval-results`, `s3://eval-artifacts/eval-nsynth-e2e/run-1`, `git_sha`, and `test/param_mse` — the full log → bytes path the entrypoint exercises. (Script asserts each token; full source pasted below.)

- [x] **Adversarial — best-effort logging never aborts a completed eval, and the no-op paths hold** *(Level 2 — the dedicated tests exercise the swallow + no-op branches against a `MagicMock(spec=WandbLogger)`)*
  ```bash
  $ uv run pytest tests/test_eval_artifact.py -v 2>&1 | grep -E "PASSED|FAILED"
  ... test_log_eval_results_artifact_swallows_logging_error PASSED
  ... test_log_eval_results_artifact_noop_when_upload_uri_unset PASSED
  ... test_log_eval_results_artifact_noop_when_no_wandb_logger PASSED
  ... test_eval_summary_metrics_drops_non_scalar_values PASSED
  ... test_eval_summary_metrics_coerces_scalar_tensors_to_floats PASSED
  (10 passed)
  ```
  A failing `log_artifact` is warned and swallowed; a null upload URI and a logger list with no `WandbLogger` both log nothing; non-scalar metrics are dropped and 0-d tensors coerced to floats.

- [ ] **`main()` gates the artifact log on global-zero and only fires when an upload URI is set** *(Level 6 — SKIP: exercising the DDP rank gate end-to-end needs a multi-device run not available here. The wiring is confirmed by source inspection.)*
  ```bash
  $ git show HEAD:src/synth_setter/cli/eval.py | sed -n '465,474p'
      is_global_zero = object_dict["trainer"].is_global_zero
      _maybe_upload_output_dir(cfg, is_global_zero)
      if is_global_zero:
          _log_eval_results_artifact(
              object_dict["logger"], cfg,
              cfg.evaluation.get("upload_output_dir_uri"),
              metric_dict, _get_git_sha(),
          )
  ```
  To verify the rank gate behaviorally, run the eval CLI under DDP (`trainer=ddp`, >1 device) with `evaluation.upload_output_dir_uri` set and a WandbLogger, then confirm exactly one `eval-results` artifact version is logged (non-zero ranks must not log).

<details><summary>E2E script (<code>/tmp/verify_eval_e2e.py</code>)</summary>

```python
import glob, os
from pathlib import Path
import wandb
from lightning.pytorch.loggers.wandb import WandbLogger
from omegaconf import OmegaConf
from synth_setter.cli.eval import _log_eval_results_artifact
from tests.helpers.wandb_offline import read_run_binary

tmp = Path(os.environ["E2E_TMP"])
for key in [k for k in os.environ if k.startswith("WANDB_")]:
    os.environ.pop(key, None)
os.environ["WANDB_MODE"] = "offline"
os.environ["WANDB_DATA_DIR"] = str(tmp / "wandb-data")
wandb.teardown()
run_id = "eval-e2e-20260606T000000000Z"
logger = WandbLogger(save_dir=str(tmp), offline=True, id=run_id, project="synth-setter", job_type="evaluation")
logger.experiment
cfg = OmegaConf.create({"task_name": "nsynth-e2e"})
upload_uri = "r2://eval-artifacts/eval-nsynth-e2e/run-1"
_log_eval_results_artifact([logger], cfg, upload_uri, {"test/param_mse": 0.125}, "a" * 40)
logger.experiment.finish()
offline_dirs = list((tmp / "wandb").glob(f"offline-run-*-{run_id}"))
assert len(offline_dirs) == 1, offline_dirs
binaries = glob.glob(str(offline_dirs[0] / "run-*.wandb"))
assert len(binaries) == 1, binaries
art_name = b"eval-nsynth-e2e"; s3_ref = b"s3://eval-artifacts/eval-nsynth-e2e/run-1"
payload = read_run_binary(Path(binaries[0]), until=lambda d: art_name in d and s3_ref in d)
assert art_name in payload and b"eval-results" in payload and s3_ref in payload
assert b"git_sha" in payload and b"test/param_mse" in payload
print("E2E PASS: eval-results artifact, type, s3 reference, and metadata all recorded")
```
</details>

File: src/synth_setter/cli/eval.py [hunk 1]
```diff
@@ -373,7 +462,16 @@ def main(cfg: DictConfig) -> None:
 
     metric_dict, object_dict = evaluate(cfg)
     _dump_metric_dict(metric_dict, Path(cfg.paths.output_dir))
-    _maybe_upload_output_dir(cfg, object_dict["trainer"].is_global_zero)
+    is_global_zero = object_dict["trainer"].is_global_zero
+    _maybe_upload_output_dir(cfg, is_global_zero)
+    if is_global_zero:
```
[the remark this answers is not in the document]
Comment (ktinubu) [reply src/synth_setter/cli/eval.py]: Fixed in 48879278: the metrics dump, R2 mirror, and eval-results artifact logging now run at the tail of `evaluate()` — *before* `@task_wrapper`'s `finally` calls `wandb.finish()` — so the artifact attaches to the open eval run. `main()` just delegates to `evaluate()`. git_sha is only computed on the global-zero + configured-upload-URI path.

File: tests/test_eval_artifact.py [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+    assert summary == {"scalar": 3.0}
+
+
+def _wandb_logger_mock() -> MagicMock:
+    """Return a ``WandbLogger`` mock whose ``.experiment.log_artifact`` is observable.
+
+    :returns: A ``MagicMock(spec=WandbLogger)`` so ``isinstance`` narrows it.
+    """
+    return MagicMock(spec=WandbLogger)
+
```
[the remark this answers is not in the document]
Comment (ktinubu) [reply tests/test_eval_artifact.py:83]: `MagicMock(spec=WandbLogger)` *does* satisfy `isinstance(m, WandbLogger)` — `spec=` sets the mock's `__class__` to the spec type, so the `isinstance` check in `_log_eval_results_artifact` passes and the test exercises the logging path (verified: `isinstance(MagicMock(spec=WandbLogger), WandbLogger) == True`, and the test is green in CI). No change needed.

File: src/synth_setter/cli/eval.py [hunk 3]
```diff
@@ -293,6 +295,17 @@ def evaluate(cfg: DictConfig) -> tuple[dict[str, Any], dict[str, Any]]:
...
+    # Persist + publish results here, not after evaluate() returns: @task_wrapper's
+    # finally closes the W&B run on return, so the eval-results artifact has to be
+    # logged while the run is still open or it would attach to nothing.
+    _dump_metric_dict(metric_dict, Path(cfg.paths.output_dir))
+    _maybe_upload_output_dir(cfg, trainer.is_global_zero)
+    upload_uri = cfg.evaluation.get("upload_output_dir_uri")
+    # _get_git_sha() shells out, so only invoke it on the path that actually logs
+    # the artifact (global-zero with a configured R2 prefix).
+    if trainer.is_global_zero and upload_uri:
+        _log_eval_results_artifact(logger, cfg, upload_uri, metric_dict, _get_git_sha())
```
[the remark this answers is not in the document]
Comment (ktinubu) [reply src/synth_setter/cli/eval.py]: Fixed in f1c566a8 — gated `_dump_metric_dict` on `trainer.is_global_zero` so only the global-zero rank writes `metrics/metrics.json`, matching the existing rank gate on the upload + artifact log. Non-zero ranks no longer race on the shared output_dir.

File: src/synth_setter/pipeline/r2_io.py [hunk 4]
```diff
@@ -178,6 +178,22 @@ def to_rclone_path(r2_uri: str) -> str:
 _to_rclone_path = to_rclone_path
 
 
+def to_s3_uri(r2_uri: str) -> str:
+    """Rewrite an ``r2://`` URI to the ``s3://`` scheme W&B references record.
+
+    R2 exposes an S3-compatible API; only the scheme differs, so the
```
[the remark this answers is not in the document]
Comment (ktinubu) [reply src/synth_setter/pipeline/r2_io.py:185]: Fixed in f1c566a8 — added `"to_s3_uri"` to `__all__` (alphabetically after `to_rclone_path`) so the public helper is exported consistently.

File: src/synth_setter/cli/eval.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (ktinubu) [reply src/synth_setter/cli/eval.py]: Fixed in f1c566a8: `_dump_metric_dict` is now gated behind `if trainer.is_global_zero:` (alongside the R2 mirror and artifact log), so under DDP non-zero ranks no longer race writes to the shared `metrics/metrics.json`.

File: src/synth_setter/pipeline/r2_io.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (ktinubu) [reply src/synth_setter/pipeline/r2_io.py:185]: Fixed in f1c566a8: `"to_s3_uri"` is now in `r2_io.__all__`, alongside `to_rclone_path`, so the public surface is consistent.

File: src/synth_setter/cli/eval.py [hunk 5]
```diff
@@ -12,9 +12,11 @@
 import wandb
 from lightning import Callback, LightningDataModule, LightningModule, Trainer
 from lightning.pytorch.loggers import Logger
+from lightning.pytorch.loggers.wandb import WandbLogger
 from omegaconf import DictConfig
 
 from synth_setter.pipeline import r2_io
+from synth_setter.pipeline.schemas.spec import _get_git_sha
 from synth_setter.resources import as_file, vst_headless_wrapper
```
[the remark this answers is not in the document]
Comment (ktinubu) [reply src/synth_setter/cli/eval.py:20]: Deferred, not adopted in this PR. Promoting `_get_git_sha` to a public helper is a cross-module rename that ripples well beyond the eval-artifact feature: `DatasetSpec.default_factory` calls it, and eight tests monkeypatch the symbol via its private path (`...schemas.spec._get_git_sha`) plus the documented `_GIT_UNAVAILABLE_SENTINEL` contract. Per repo convention a rename of that reach goes through `/tdd-refactor` (exhaustive reference discovery + contract pinning) as its own `refactor:` PR, not bundled into a feature PR. The current cross-module import reuses the single existing implementation rather than duplicating the git-SHA logic, and the symbol's behavior is pinned by its docstring + sentinel — so it is a stable de-facto API. Tracking the public-promotion as separate follow-up work.
